i18n_generate.mll

```diff
@@ -305,28 +403,71 @@ let _ =
       let x = normalize_type ?primary_module x in
       assert (List.mem x variants) ;
       x in
-  let lexbuf = Lexing.from_channel in_chan in
-  (try
-     let key_values = parse_lines variants [] lexbuf in
+  
      let output = Format.formatter_of_out_channel out_chan in
-     if primary_module = None && not (!external_type) then
-       ( print_type output ~variants
-       ; print_string_of_language output ~variants ~strings
-       ; print_language_of_string output ~variants ~strings
-       ; print_guess_language_of_string output) ;
-     print_list_of_languages output ~variants ;
-     print_header output ?primary_module ~default_language () ;
-     Format.pp_print_string output "[%%shared\n" ;
-     Format.fprintf output "module Tr = struct\n" ;
-     print_module_body print_expr_html output key_values ;
-     Format.fprintf output "\nmodule S = struct\n" ;
-     print_module_body print_expr_string output key_values ;
-     Format.fprintf output "\nend\n" ;
-     Format.fprintf output "end\n" ;
-     Format.pp_print_string output "]\n"
+     if !header then (
+       if !eliom_generation then
+         (print_type_eliom output ~variants
+         ; print_string_of_language_eliom output ~variants ~strings
+         ; print_language_of_string_eliom output ~variants ~strings
+         ; print_guess_language_of_string_eliom output)
+       else
+         (Format.fprintf output "open Tyxml.Html\n"
+         ; print_type output ~variants
+         ; print_string_of_language output ~variants ~strings
+         ; print_language_of_string output ~variants ~strings
+         ; print_guess_language_of_string output 
+         ; print_list_of_languages output ~variants
+         ; print_header output ?primary_module ~default_language () )
+     )
+     else ( 
+       let in_chan =
+         match !input_file with
+         | "-" -> stdin
+         | file -> open_in file in
+       let lexbuf = Lexing.from_channel in_chan in
+       try
+       let key_values = parse_lines variants [] lexbuf in
+       if !eliom_generation then 
```

Comment:
this if looks like it could be refactored because both branches share a lot of code